### DISCUSSION TITLE
Strategies for debugging with tar_make_future() ?

### DISCUSSION THREAD

**Discussion by JakeRuss** [ropensci/targets#678] [Help] [answered]
Hi Will (and other targets folks), 

Is there a way capture why a target failed via `tar_make_future()`? I'm getting error messages about "callr subprocess failures" but I can't seem to find out what triggered the failure. 

When I use `plan(callr)` I get 

```
x error branch model_scores_1bff76c2
• end pipeline
Error : _future_ callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
In addition: Warning message:
You're using a user token but also specified the following parameter(s): username, NA. These arguments will have no effect.
This warning is displayed once per session. 
 Error: callr subprocess failed: _future_ callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
Visit https://books.ropensci.org/targets/debugging.html for debugging advice.
Run `rlang::last_error()` to see where the error occurred. 
> rlang::last_error()
<error/tar_condition_run>
callr subprocess failed: _future_ callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
Visit https://books.ropensci.org/targets/debugging.html for debugging advice.
Backtrace:
 1. global::run_targets_run_models()
 2. targets::tar_make_future(...) ~/rnc-2022-cycle/projects/state/va/01_september_abev/project_information.R:142:2
 3. targets:::callr_outer(...)
 4. base::tryCatch(...)
 5. base:::tryCatchList(expr, classes, parentenv, handlers)
 6. base:::tryCatchOne(expr, names, parentenv, handlers[[1L]])
 7. value[[3L]](cond)
 8. targets::tar_throw_run(...)
Run `rlang::last_trace()` to see the full context.
```
I don't really see anything useful in that, so I tried switching to `plan(multisession)` and I get, 
```
start branch model_scores_9a233c00
✓ skip pipeline
Error in unserialize(node$con) : 
  Failed to retrieve the value of MultisessionFuture (model_scores_c4981d99) from cluster RichSOCKnode #1 (PID 2949 on localhost 'localhost'). The reason reported was 'error reading from connection'. Post-mortem diagnostic: No process exists with this PID, i.e. the localhost worker is no longer alive.
In addition: Warning message:
You're using a user token but also specified the following parameter(s): username, NA. These arguments will have no effect.
This warning is displayed once per session. 
Error: callr subprocess failed: Failed to retrieve the value of MultisessionFuture (model_scores_c4981d99) from cluster RichSOCKnode #1 (PID 2949 on localhost 'localhost'). The reason reported was 'error reading from connection'. Post-mortem diagnostic: No process exists with this PID, i.e. the localhost worker is no longer alive.
Visit https://books.ropensci.org/targets/debugging.html for debugging advice.
Run `rlang::last_error()` to see where the error occurred.
> rlang::last_error()
<error/tar_condition_run>
callr subprocess failed: Failed to retrieve the value of MultisessionFuture (model_scores_c4981d99) from cluster RichSOCKnode #1 (PID 2949 on localhost 'localhost'). The reason reported was 'error reading from connection'. Post-mortem diagnostic: No process exists with this PID, i.e. the localhost worker is no longer alive.
Visit https://books.ropensci.org/targets/debugging.html for debugging advice.
Backtrace:
 1. global::run_targets_run_models()
 2. targets::tar_make_future(...)
 3. targets:::callr_outer(...)
 4. base::tryCatch(...)
 5. base:::tryCatchList(expr, classes, parentenv, handlers)
 6. base:::tryCatchOne(expr, names, parentenv, handlers[[1L]])
 7. value[[3L]](cond)
 8. targets::tar_throw_run(...)
Run `rlang::last_trace()` to see the full context.
```
Which also isn't much help to me. 

One other thing I tried was going to future directly, where I loaded my target's inputs and tried to run the function that way, naturally, it worked just fine, so I'm missing something when I move to more than one worker. 

```
library(future)
library(targets)

options(future.globals.maxSize = 3670016000)
plan(callr)

library(tidyverse)
library(tidymodels)

tar_load(wflows_final_118b626a, store = local_stage3_dir)
tar_load(baked_data, store = local_stage3_dir)

x <- "replacement_name"

f <- future( workflows::extract_fit_parsnip(wflows_final_118b626a) %>%
               parsnip::augment(new_data = baked_data, type = "prob") %>%
               dplyr::select(id, var1, var2, var3, !! x := .pred_1) %>%
               tidyr::pivot_longer(cols = matches(x), names_to = "model_name", values_to = "score"))

value(f)
```

My whole pipeline works sequentially, but one step fails when I move to tar_make_future, and a reprex would be complicated to set up. So instead of asking for help directly on my problem, I'm asking how to better help myself?

**Comment by wlandau** [comment]
Sorry, it's hard to tell what could be happening. You are already peeling back layers, e.g. running it without `targets`. More things you could try include `tar_make_future(callr_function = NULL)`, commenting out targets until you find the smallest number that produce the error, removing the `tidymodels` content and see if the error persists, etc. How exactly are you setting up your `future` plan in `_targets.R`?

**Reply by JakeRuss** [reply]
Thanks Will. I know asking for help without code is risky business. What does `tar_make_future(callr_function = NULL)` do in the parallel context?

My _target.R:
```
library(targets)
library(tarchetypes)
library(here)
library(dplyr)
library(tidyverse)
library(slackr)
library(future)
library(future.callr)

options(future.globals.maxSize = 3800016000)
#options(future.globals.onReference = "error")
#future::plan(multisession)
future::plan(callr)

source("functions.R") 

# Set Target Default Options  -------------------------------------------------

tar_option_set(packages  = c("tidyverse", "tidymodels"),
               tidy_eval = TRUE,
               memory    = "transient",
               storage   = "worker",
               retrieval = "worker",
               resources = tar_resources(
                 aws = tar_resources_aws(
                   bucket = project_s3_bucket,
                   prefix = paste0(project_s3_prefix, "/week", project_round_number, "/stage3_storage")))
)

list(targets... )
```

**Reply by JakeRuss** [reply]
I tried running with `tar_make_future(callr_function = NULL)` and that created new errors for "function not found" errors for everything in `functions.R`

**Reply by wlandau** [reply]
> I tried running with tar_make_future(callr_function = NULL) and that created new errors for "function not found" errors for everything in functions.R

That's odd, not sure what could be wrong. Would need to run it myself.

A few more things you could try independently:

* `plan(sequential)`
* `storage` and `retrieval` equal to `"main"`.
* local storage instead of AWS

Might give us more clues.

**Reply by JakeRuss** [reply]
Many thanks Will. I'm going to back away from this problem for today, and clear my head, because I feel like I've run myself in circles over this one. I'll pick back up on it tomorrow and systematically try all of your suggestions you've listed. If none of that works, I'll attempt to make a sharable example.

**Reply by JakeRuss** [reply]
@wlandau I tried the `plan(sequential)` suggestion and the error messages went away. But the more surprising result (to me) was that with `workers = 2` I was still getting two branches being built at the same time. Now I'm slightly confused, I thought `plan(sequential)` meant no parallelization. I'm guessing the tasks were split across two workers and then run sequentially?

**Reply by JakeRuss** [reply]
I also created a shareable example of our pipeline, but it ran just fine in the `plan(callr)` version (naturally). So I'm going to add AWS back in and see if I can break it.

**Reply by JakeRuss** [reply]
Hey @wlandau, I know you have other things going on, but I finally figured it out and wanted to report back. Turns out the size `baked_data` object was the culprit. When I trimmed that down to one-third its size, all of the errors go away and the pipeline works with all of the various plan options. Originally I thought increasing `options(future.globals.maxSize = )` would help, but I guess I have to figure out how to break the prediction step into smaller parts or figure out how to give the workers more resources. 

I did create a shareable example of the pipeline and if you'd like I could contribute it somehow to the targets docs. It runs multiple tidymodels random forest fits, eg. a model for outcome 1, a model for outcome 2, which is not something tidymodels is good at directly--right now those tools are mostly focused on a single outcome and running multiple versions of that model.

**Comment by wlandau** [comment, marked as answer]
> Hey @wlandau, I know you have other things going on, but I finally figured it out and wanted to report back. Turns out the size baked_data object was the culprit. When I trimmed that down to one-third its size, all of the errors go away and the pipeline works with all of the various plan options. Originally I thought increasing options(future.globals.maxSize = ) would help, but I guess I have to figure out how to break the prediction step into smaller parts or figure out how to give the workers more resources.

Thanks for figuring that out. You may have better luck with `tar_option_set(storage = "worker", retrieval = "worker")`. That way, the parallel workers store and retrieve the data, and there is very little data to serialize and ship over a network.

> I did create a shareable example of the pipeline and if you'd like I could contribute it somehow to the targets docs. It runs multiple tidymodels random forest fits, eg. a model for outcome 1, a model for outcome 2, which is not something tidymodels is good at directly--right now those tools are mostly focused on a single outcome and running multiple versions of that model.

Sounds great, `targets` + `tidymodels` deserves exploration. If you put it in a repo, we could list it under https://docs.ropensci.org/targets/#example-projects. Could also be nice to have an RStudio Cloud space just like the other linked examples.

**Reply by JakeRuss** [reply]
I'll mark this as resolved and we can continue a discussion in that repo I'll start. Thanks
